상품 상세 페이지 메인 영역

youzysu opened the pull request:
## Issues
- #49 

## What is this PR? 👓
- 상품 상세 페이지 UI
- 상품 정보 상세 조회
- 상품 상태 변경 기능
![Screen Recording 2023-09-13 at 20 19 27](https://github.com/masters2023-project-team05-second-hand/second-hand-max-fe/assets/111998760/b81b6ce7-4fbf-4fa9-8716-bf07c4897d30)

## To reviewers 👋
- 상품 상세 이미지 영역 swiper 라이브러리 활용하여 구현했어요.
- Product에서 재사용하는 type 분리했어요.
- 상품 상태는 기존 Modal을 활용해서 상태 변경 모달에서 변경하도록 구현했어요. (재사용을 위해 기존 모달 스타일 관련 수정이 있었습니다)
- 기타 API 변경사항 반영하면서 리팩토링했습니다!
- 공통 스타일 컴포넌트 생성 및 Page 레이아웃 수정 등 스타일 작업했습니다.

youzysu committed:
refactor: 상품 상세 조회 API 응답 객체 data 반환 및 반영

youzysu committed:
refactor: TopBar 컴포넌트 리팩토링

- z-index 추가
- title 빈 element 반환하지 않도록 수정

youzysu committed:
refactor: useScroll dependencies 빈배열 추가, 타입 수정

youzysu committed:
feat #49: 상품 상세 페이지 이미지 영역

youzysu committed:
feat #49: 상품 상세 정보 response 변경사항 반영, 판매자 정보 컴포넌트 UI

youzysu committed:
feat #49: 상품 상세 페이지 상품 설명 본문 UI

youzysu committed:
refactor: API 변경사항 전반 반영

- 유저 동네 정보 isLastVisited 필드 삭제
- 상태 목록 조회 API, 상품 상태 변경 API 추가
- Modal 재사용을 위한 스타일 수정 및 props 네이밍 수정 (headerProps -> header)

youzysu committed:
feat #49: 상품 상세 페이지 상품 상태 변경 기능

- 상품 상태 버튼 및 상태 변경 모달 UI
- 상품 상태 목록 조회, mutateProductStatus 후 상품 상세 다시 조회하기(invalidateQueries 활용)

youzysu committed:
fix: Member 타입 재수정, useMutation 구조 분해 할당 리팩토링

youzysu committed:
design: BottomBar 공통 스타일 컴포넌트 생성, Page 레이아웃 수정

- 페이지 끝까지 스크롤하기 위해 Page height 조정

youzysu committed:
UI #49: 제품 상세 페이지 하단바 UI

youzysu committed:
design: accentPrimary filter 값 수정

youzysu committed:
design: Toast 스타일 수정

youzysu committed:
fix: 상품 수정 페이지 상품 가격 null 체크

youzysu committed:
fix: MenuIndicator import

youzysu commented on fe/src/styles/common.ts:
```diff
@@ -41,3 +40,17 @@ export const Dim = styled.div`
   background-color: ${({ theme: { color } }) => color.neutralOverlay};
   z-index: -1;
 `;
```
상품 상세 페이지에서 BottomBar 윗부분까지 제품 상세 설명이 다 보이게 하기 위해서 Page의 height에서 하단 높이를 빼줬는데, 혹시 다른 방법이 있을까요?

youzysu committed:
refactor: 상품 상태 변경 useMutateProductStatus hook 생성

youzysu commented on fe/src/api/product/queries.ts:
```diff
@@ -20,9 +32,47 @@ export const useCategoryQuery = () => {
+  });
+};
+
+export const useMutateProductStatus = ({
+  onSettled,
+  productId,
+}: {
+  onSettled: () => void;
+  productId: number;
+}) => {
+  const queryClient = useQueryClient();
+  const { toast } = useToast();
+
+  return useMutation(patchProductStatus, {
+    onSuccess: () => {
+      toast({
+        type: "success",
+        title: "상태 변경 완료",
+        message: "상태 변경이 완료되었습니다.",
+      });
+      queryClient.invalidateQueries(["getProductDetail", productId]);
+    },
+    onError: () => {
+      toast({
+        type: "error",
+        title: "상태 변경 실패",
+        message: "상태 변경에 실패했습니다. 잠시 후 다시 시도해주세요.",
+      });
+    },
+    onSettled,
+  });
+};
```
분리한 상품 상태 변경 hook

youzysu committed:
feat #49: 상품 상세 페이지 이전 페이지 이동 기능

- 상품 상세 페이지 이전 라우트 위치에 따라 이전 화면/메인 화면으로 이동 분기
- ROUTE_PATH 상수 리팩토링 (하드코딩 수정)

aaaz425 commented on fe/src/components/ProductDetail/ProductImageList.tsx:
```diff
@@ -0,0 +1,80 @@
+}) {
+  return (
+    <StyledProductImageList>
+      <Swiper
+        pagination={{
+          type: "fraction",
+        }}
+        className="product-image-swiper"
+        modules={[Pagination]}>
+        {productImages.map((productImage) => (
+          <SwiperSlide key={productImage.id}>
+            <ProductImage
+              key={productImage.id}
+              src={productImage.url}
+              alt={`${productImage.id} 상품 이미지`}
+            />
+          </SwiperSlide>
+        ))}
+      </Swiper>
```
👍

aaaz425 commented on fe/src/styles/common.ts:
```diff
@@ -1,16 +1,15 @@
+// 참고: scroll을 Page에서 제어하고 있어서 min-height 하단바의 높이만큼 빼줘야 끝까지 스크롤 가능
 export const Page = styled.div`
   position: relative;
   margin: 0 auto;
-  max-width: ${WIDTH.page}px;
-  min-height: 100vh;
-  height: 100vh;
+  width: ${WIDTH.page}px;
+  height: calc(100vh - ${HEIGHT.navigationBar}px);
```
👍👍 원래 하단바 높이만큼 빈 공간 생겨서 수정하신거죠??

aaaz425 commented on fe/src/utils/time.ts:
```diff
@@ -0,0 +1,41 @@
+];
+
+export const convertPastTimestamp = (timestamp: string) => {
+  const startDate = new Date(timestamp);
+
+  if (isNaN(startDate.getTime())) {
+    throw new Error("Invalid timestamp");
+  }
+
+  const currDate = new Date();
+  const diffMs = currDate.getTime() - startDate.getTime();
+
+  if (diffMs < 0) throw Error("timestamp cannot be in the future.");
+
+  for (const { unit, threshold, divisor } of timeUnits) {
+    if (diffMs < threshold) {
+      const value = Math.floor(diffMs / divisor);
+      return threshold === minuteInMs ? `${unit}` : `${value}${unit}`;
+    }
+  }
+
+  const year = startDate.getFullYear();
+  const month = startDate.getMonth();
+  const date = startDate.getDate();
+
+  return `${year}/${month + 1}/${date}`;
```
예외 처리도 좋고, 1년 이상 지난 시간은 풀 타임 데이터 보여줄 수 있어서 더 좋네요
조이껄로 하시죠~

youzysu committed:
refactor: 상품 상태 변경 hook 리팩토링 및 product path 상수 수정

- 상품 상태 변경 후 invalidateQueries key 인자로 받도록 hook 수정
- useProductDetailQuery productId 타입 string으로 변경, 해당 route params에 productId 반드시 존재하므로 ! 사용
- product path 중복 제거 및 api에서 productId 추가하도록 변경

youzysu committed:
fix: error 해결 및 리팩토링
